src/ess/reduce/time_of_flight/eto_to_tof.py

```diff
@@ -284,38 +284,89 @@ def compute_tof_lookup_table(
-    base_grid = np.arange(float(lookup.sizes["pulse"]))
-    return RegularGridInterpolator(
-        (
-            np.sort(np.concatenate([base_grid - 0.25, base_grid + 0.25])),
-            lookup.coords["distance"].to(unit=distance_unit, copy=False).values,
-            lookup.coords["event_time_offset"].to(unit=time_unit, copy=False).values,
-        ),
-        np.repeat(lookup.data.to(unit=time_unit, copy=False).values, 2, axis=0),
-        method="linear",
-        bounds_error=False,
-        fill_value=np.nan,
-    )
+class TofInterpolator:
+    def __init__(self, lookup: sc.DataArray, distance_unit: str, time_unit: str):
+        from scipy.interpolate import RegularGridInterpolator
+
+        # TODO: to make use of multi-threading, we could write our own interpolator.
+        # This should be simple enough as we are making the bins linspace, so computing
+        # bin indices is fast.
+
+        self._distance_unit = distance_unit
+        self._time_unit = time_unit
+
+        # In the pulse dimension, it could be that for a given event_time_offset and
+        # distance, a tof value is finite in one pulse and NaN in the other.
+        # When using the bilinear interpolation, even if the value of the requested
+        # point is exactly 0 or 1 (in the case of pulse_stride=2), the interpolator
+        # will still use all 4 corners surrounding the point. This means that if one of
+        # the corners is NaN, the result will be NaN.
+        # Here, we use a trick where we duplicate the lookup values in the 'pulse'
+        # dimension so that the interpolator has values on bin edges for that dimension.
+        # The interpolator raises an error if axes coordinates are not strictly
+        # monotonic, so we cannot use e.g. [-0.5, 0.5, 0.5, 1.5] in the case of
+        # pulse_stride=2. Instead we use [-0.25, 0.25, 0.75, 1.25].
+        base_grid = np.arange(float(lookup.sizes["pulse"]))
+        self._interpolator = RegularGridInterpolator(
+            (
+                np.sort(np.concatenate([base_grid - 0.25, base_grid + 0.25])),
+                lookup.coords["distance"].to(unit=distance_unit, copy=False).values,
+                lookup.coords["event_time_offset"]
+                .to(unit=self._time_unit, copy=False)
+                .values,
+            ),
+            np.repeat(
+                lookup.data.to(unit=self._time_unit, copy=False).values, 2, axis=0
+            ),
+            method="linear",
+            bounds_error=False,
+            fill_value=np.nan,
+        )
+
+    def __call__(
+        self,
+        pulse_index: sc.Variable,
+        ltotal: sc.Variable,
+        event_time_offset: sc.Variable,
+    ) -> sc.Variable:
+        if pulse_index.unit != sc.units.dimensionless:
+            raise ValueError(
+                f"pulse_index must be dimensionless, but got unit: {pulse_index.unit}."
+            )
+        if ltotal.unit != self._distance_unit:
+            raise ValueError(
```

Comment:
```suggestion
            raise sc.UnitError(
```
?
And the same for the other cases.